src/libs/data-utils.js

```diff
@@ -340,3 +356,231 @@ export const EntityUploader = class EntityUploader extends Component {
+        ])
+      ]) :
+      h(Fragment, [
+        div({ style: { marginBottom: '1rem' } }, [
+          fieldset({ style: { border: 'none', margin: 0, padding: 0 } }, [
+            legend({ style: { marginBottom: '0.5rem' } }, [isList ? 'List item type:' : 'Attribute type:']),
+            h(Fragment, _.map(({ type, tooltip }) => h(TooltipTrigger, {
+              content: tooltip
+            }, [
+              span({ style: { marginRight: '1.2rem' } }, [h(RadioButton, {
+                text: _.startCase(type),
+                name: 'edit-type',
+                checked: editType === type,
+                onChange: () => setEditType(type),
+                labelStyle: { paddingLeft: '0.5rem' }
+              })])
+            ]), [
+              { type: 'string' },
+              { type: 'reference', tooltip: 'A link to another entity' },
+              { type: 'number' },
+              { type: 'boolean' }
+            ]))
+          ]),
+          editType === 'reference' && div({ style: { marginTop: '0.5rem' } }, [
+            div({ style: { marginBottom: '0.5rem' } }, 'Referenced entity type:'),
+            h(Select, {
+              value: linkedEntityType,
+              options: entityTypes,
+              onChange: ({ value }) => setLinkedEntityType(value)
+            })
+          ])
+        ]),
+        div({ style: { marginBottom: '0.5rem' } }, [
+          h(LabeledCheckbox, {
+            checked: isList,
+            onChange: willBeList => setNewValue(willBeList ? [newValue] : newValue[0])
+          }, [span({ style: { marginLeft: '0.5rem' } }, ['Attribute is a list'])])
+        ]),
+        isList ?
+          div({ style: { marginTop: '1.5rem' } }, _.map(([i, value]) => div({
+            style: { display: 'flex', alignItems: 'center', marginBottom: '0.5rem' }
+          }, [
+            makeInput({
+              'aria-label': `List value ${i + 1}`,
+              autoFocus: true,
+              placeholder: 'Enter a value',
+              value,
+              onChange: v => setNewValue([...newValue.slice(0, i), v, ...newValue.slice(i + 1)])
+            }),
+            h(Link, {
+              'aria-label': `Remove list value ${i + 1}`,
+              disabled: newValue.length === 1,
+              onClick: () => setNewValue([...newValue.slice(0, i), ...newValue.slice(i + 1)]),
+              style: { marginLeft: '0.5rem' }
+            }, [
+              icon('times', { size: 20 })
+            ])
+          ]), Utils.toIndexPairs(newValue))) :
+          div({ style: { marginTop: '1.5rem' } }, [
+            makeInput({
+              'aria-label': 'New value',
+              autoFocus: true,
+              placeholder: 'Enter a value',
+              value: newValue,
+              onChange: setNewValue
+            })
+          ]),
+        isList && h(Link, {
+          style: { display: 'block', marginTop: '1rem' },
+          onClick: () => setNewValue(Utils.append(''))
```

Comment:
TOL: The handling of data types in state is currently a bit loose, and it creates some surprising interactions. For example, after switching types, the underlying data doesn't change until you interact with the new input control. That can lead to arrays of mixed types, and edge cases like the string`'false'` mapping to the boolean `true`. I think everything sorts itself out before sending to the server, but it seems like a risky area for bugs. I wonder if we should be more formal about managing the internal state.

Reply:
Take a look at what I did.